setuptools/dist.py

```diff
@@ -438,35 +440,73 @@ def _finalize_license_files(self) -> None:
         """Compute names of all license files which should be included."""
         license_files: list[str] | None = self.metadata.license_files
         patterns = license_files or []
+        skip_pattern_validation = False
 
         license_file: str | None = self.metadata.license_file
         if license_file and license_file not in patterns:
             patterns.append(license_file)
+            skip_pattern_validation = True
 
         if license_files is None and license_file is None:
             # Default patterns match the ones wheel uses
             # See https://wheel.readthedocs.io/en/stable/user_guide.html
             # -> 'Including license files in the generated wheel file'
             patterns = ['LICEN[CS]E*', 'COPYING*', 'NOTICE*', 'AUTHORS*']
+            skip_pattern_validation = True
+
+        matched_files = []
+        if skip_pattern_validation is True:
+            for pattern in patterns:
+                matched_files.extend(self._expand_pattern(pattern))
+        else:
+            for pattern in patterns:
+                matched_files.extend(self._validate_and_expand_pattern(pattern))
 
         self.metadata.license_files = list(
             map(
                 lambda path: path.replace(os.sep, "/"),
-                unique_everseen(self._expand_patterns(patterns)),
+                unique_everseen(matched_files),
             )
         )
 
+    def _validate_and_expand_pattern(self, pattern):
+        """Validate license file patterns according to the PyPA specifications.
+        https://packaging.python.org/en/latest/specifications/pyproject-toml/#license-files
+        """
+        if ".." in pattern:
+            raise InvalidConfigError(
+                f"License file pattern '{pattern}' cannot contain '..'"
+            )
+        if pattern.startswith((os.sep, "/")) or ":\\" in pattern:
+            raise InvalidConfigError(
+                f"License file pattern '{pattern}' should be relative and "
+                "must not start with '/'"
+            )
+        if _license_files_allowed_chars.match(pattern) is None:
+            raise InvalidConfigError(
+                f"License file pattern '{pattern}' contains invalid "
+                "characters. "
+                "https://packaging.python.org/en/latest/specifications/pyproject-toml/#license-files"
+            )
```

Comment:
This is similar to the validation pattern I added in `flit` a few weeks ago. https://github.com/pypa/flit/pull/705